File: backend/app/schemas/candles.py
from pydantic import BaseModel
from typing import Union


class Candle(BaseModel):
    # intraday candles use a UNIX timestamp (int), daily candles use "YYYY-MM-DD" (str)
    time: Union[int, str]
    open: float
    high: float
    low: float
    close: float
    volume: float


class CandleResponse(BaseModel):
    ticker: str
    range: str
    interval: str
    candles: list[Candle]


class StockProfile(BaseModel):
    ticker: str
    company_name: str | None = None
    currency: str | None = None
    exchange: str | None = None
    market_cap: float | None = None
    pe_ratio: float | None = None
    eps: float | None = None
    fifty_two_week_high: float | None = None
    fifty_two_week_low: float | None = None
    day_high: float | None = None
    day_low: float | None = None
    volume: int | None = None
    avg_volume: int | None = None
    dividend_yield: float | None = None
    sector: str | None = None

File: frontend/src/types/candles.ts
export interface Candle {
  // intraday = UNIX timestamp (number), daily = "YYYY-MM-DD" (string)
  time: number | string;
  open: number;
  high: number;
  low: number;
  close: number;
  volume: number;
}

export interface CandleResponse {
  ticker: string;
  range: string;
  interval: string;
  candles: Candle[];
}

export interface StockProfile {
  ticker: string;
  company_name: string | null;
  currency: string | null;
  exchange: string | null;
  market_cap: number | null;
  pe_ratio: number | null;
  eps: number | null;
  fifty_two_week_high: number | null;
  fifty_two_week_low: number | null;
  day_high: number | null;
  day_low: number | null;
  volume: number | null;
  avg_volume: number | null;
  dividend_yield: number | null;
  sector: string | null;
}

export interface Earnings {
  ticker: string;
  earnings_date: string | null;
  eps_estimate: number | null;
  eps_actual: number | null;
  eps_surprise_pct: number | null;
  revenue_actual: number | null;
  operating_income: number | null;
  gross_margin: number | null;
  operating_margin: number | null;
  last_updated: string | null;
}

export type RangeOption = '1D' | '1W' | '1M' | '1Y';
